### PR TITLE
Remove visiblity of libraries in inherited contracts.

### LINKED ISSUE #8188: Usage of libraries is inherited
Today I stumbled with a strange scenario that I did not expect to work, yet it does: contracts that inherit from parents with a `using` statement can also use the same library in that manner.

Consider the following contracts:
```
pragma solidity ^0.5.0;

library Lib {
    function foo(uint256 value) internal returns (uint256) {
        return value + 42;
    }
}

contract A {
    using Lib for uint256;
}

contract B is A {
    function bar(uint256 value) public returns (uint256) {
        return value.foo(); // Usage of Lib
    }
}
```

Note that `B` could even exist on a separate source file, which would make the code all that more cryptic - it's probably not very hard to introduce a backdoor using this mechanism.

Is this behavior intended, or a bug?

### PR DESCRIPTION
**Pull request by Marenz**
fixes #8188

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -10,6 +10,7 @@ Breaking changes:
  * Type Checker: Exponentiation and shifts of literals by non-literals will always use ``uint256`` or ``int256`` as a type.
  * Type Checker: Disallow structs and arrays in memory or calldata if they contain nested mappings.
  * Type Checker: Disallow assignments to state variables that contain nested mappings.
+ * Remove visibility of Libraries in inherited contracts.
```
Comment (chriseth) [inline Changelog.md]: `using A for B` only affects the contract it is mentioned in and not all derived contracts

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -356,8 +356,7 @@ MemberList::MemberMap Type::boundFunctions(Type const& _type, ASTNode const& _sc
 		usingForDirectives += ASTNode::filteredNodes<UsingForDirective>(sourceUnit->nodes());
 	else if (auto const* contract = dynamic_cast<ContractDefinition const*>(&_scope))
 	{
-		for (ContractDefinition const* contract: contract->annotation().linearizedBaseContracts)
-			usingForDirectives += contract->usingForDirectives();
+		usingForDirectives += contract->usingForDirectives();
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp]: Can you turn this into a single statement and remove the braces?

**Comment by chriseth** [comment]
Did you check if any changes to the documentation are needed? Please also add something to the list "how to update to 0.7.x".

File: docs/070-breaking-changes.rst [hunk 3]
```diff
@@ -18,3 +18,6 @@ This section gives detailed instructions on how to update prior code for every b
 * Change ``now`` to ``block.timestamp``.
 * Change types of right operand in shift operators to unsigned types. For example change ``x >> (256 - y)`` to
   ``x >> uint(256 - y)``.
+* Any contract that uses internal functions of a library that was made available
```
Comment (chriseth) [inline docs/070-breaking-changes.rst]: Note that you can use internal functions of a library without the `using` statement by just `L.f()`. So maybe rather say something like `if you want a "using" statement to be effective in derived contracts, add it to these contracts explicitly.`

File: docs/contracts/libraries.rst [hunk 4]
```diff
@@ -28,16 +28,17 @@ not possible to destroy a library.
...
-to library functions look just like calls to functions of explicit base
-contracts (``L.f()`` if ``L`` is the name of the library). Furthermore,
-``internal`` functions of libraries are visible in all contracts, just as
-if the library were a base contract. Of course, calls to internal functions
-use the internal calling convention, which means that all internal types
-can be passed and types :ref:`stored in memory <data-location>` will be passed by reference and not copied.
-To realize this in the EVM, code of internal library functions
-and all functions called from therein will at compile time be included in the calling
-contract, and a regular ``JUMP`` call will be used instead of a ``DELEGATECALL``.
+They will not be explicitly visible in the inheritance hierarchy, but calls to
```
Comment (chriseth) [inline docs/contracts/libraries.rst]: Please undo line break changes, this is difficult to review.

File: docs/contracts/libraries.rst [hunk 5]
```diff
@@ -30,9 +30,10 @@ not possible to destroy a library.
 Libraries can be seen as implicit base contracts of the contracts that use them.
 They will not be explicitly visible in the inheritance hierarchy, but calls
 to library functions look just like calls to functions of explicit base
-contracts (``L.f()`` if ``L`` is the name of the library). Furthermore,
-``internal`` functions of libraries are visible in all contracts, just as
```
Comment (chriseth) [inline docs/contracts/libraries.rst:34]: Just add something like ``(using qualified access, like `L.f()`)`` or something.

**Comment by hrkrshnn** [comment]
The external contracts on breaking don't work anymore because of the "constructor visibility PR." We probably will have to merge it first.

File: docs/070-breaking-changes.rst (hunk 3 above)
Comment (chriseth) [inline docs/070-breaking-changes.rst]: Shouldn't this say: `Repeat the 'using A for B' statements in all derived contract if needed.'

File: docs/contracts/libraries.rst [hunk 6]
```diff
@@ -30,9 +30,10 @@ not possible to destroy a library.
...
-if the library were a base contract. Of course, calls to internal functions
+contracts (using qualified access like ``L.f()``). However,
```
Comment (chriseth) [inline docs/contracts/libraries.rst]: I don't understand what you mean my the sentence that starts with `However,` - I would just remove it.

**Review by chriseth — APPROVED** [review]

**Review by chriseth — APPROVED** [review]

**Pull request merged**